### PR TITLE
feat(follow-list): implement backend for followers/following list

### LINKED ISSUE #271: As a user, I want to be able to see users I follow and that follows me, so that I can know audience.

### LINKED ISSUE #295: Wire backend to Followers/Following screen to fetch real data

### PR DESCRIPTION
**Pull request by abaur3**
# What Changes

This pull request introduces the backend implementation for followers/following list that the user can access by clicking on followers/following stats on their own profile. Now data is not mocked anymore, but fetched directly from Cloud.

### Firestore data model

Before this PR, profile documents in firebase only held a following array as attribute, where new followed users where being added. There were two problems with this configuration:
1. There were no followers list, so fetching followers to show in the followers list would have been impossible, and
2. Storing all following/followers into arrays is not scalable, as these lists could potentially grow infinitely.
To solve this problem, I decided to create two new sub collections within a profile document that can hold all followers/following items: the `/profiles/{uid}/followers` and `/profiles/{uid}/following` subcollections. I migrated all existing entries in the following arrays to the new sub collection, and finally I updated the `followUser` Cloud function to write follow edges to subcollections. Finally, I removed the legacy `following` field from profile documents and the `Profile` data class.

### ProfileRepository updates

- Extended the `ProfileRepository` API with get/observe methods for followers and following IDs
- Implemented the new API in `ProfileRepositoryFirebase` using shared helpers for reading and observing subcollections
- Removed all remaining mapping/initialization of the old `following` array
- Updated fake repositories to match the new API and support real follow mutations

### ViewModels & UI wiring

- Wired followers/following screens to real repository calls with proper auth checks and error handling
- Follow/unfollow now optimistically updates UI state and rolls back on failure
- `MainViewModel`: observe following IDs together with samples to keep feeds in sync and simplify cache logic
- `SearchViewModel`: observe following IDs directly instead of deriving them from full profiles
- Fixed edge cases around sign-out, anonymous users, and null auth states

### Tests

- Added repository tests covering:
  - get/observe followers & following
  - Firestore listener error handling (safe defaults)
  - Correct ID mapping and deduplication
- Extended follow list tests to cover:
  - Follow/unfollow sync edge cases across tabs
  - Error paths (no auth user, repository failures)
  - Progress and state rollback on failure
- Updated existing tests to remove all dependencies on the old `following` array

### Firestore rules

- Signed-in users can read profile documents and follower/following subcollections
- Follow edges remain writable via Cloud Functions only

### Build / CI

- Increased Gradle and test JVM heap sizes
- Disabled parallel test forks to prevent Robolectric OOMs on CI

## Notes
This PR description was polished by ChatGPT for clarity.

Closes #271  
Closes #295

### CODE REVIEW

File: app/src/main/java/com/neptune/neptune/ui/follow/FollowListViewModel.kt [hunk 1]
```diff
@@ -62,92 +62,167 @@ class FollowListViewModel(
...
-      _uiState.update {
-        it.copy(following = fakeFollowing(), isLoadingFollowing = false, errorMessage = null)
+      val uid = auth.currentUser?.uid
+      if (uid == null) {
+        _uiState.update {
+          it.copy(
+              isLoadingFollowing = false, errorMessage = "User not authenticated. Please sign in.")
+        }
+        return@launch
+      }
```
Comment (AngeB21) [inline app/src/main/java/com/neptune/neptune/ui/follow/FollowListViewModel.kt]: You could create a small helper function to avoid code duplication with `loadFollower()`

File: firebase/firestore/firestore.rules [hunk 2]
```diff
@@ -2,16 +2,48 @@ rules_version = '2';
...
+      // Allow any logged-in user to READ any profile
+      allow read: if signedIn();
+
+      // Write permissions restricted to the owner
+      allow create, update, delete: if signedIn() &&
+      (uid == request.auth.uid || (request.resource.data.diff(resource.data).affectedKeys().hasOnly(['likes'])));
+
+      // Allow reading follow edges for any signed-in user; writes are done via Cloud Functions.
+      match /followers/{docId} {
+        allow read: if signedIn();
```
Comment (AngeB21) [inline firebase/firestore/firestore.rules]: maybe you should add a line `allow write: if false;` to explicitly specify that the client should not write.

File: firebase/firestore/firestore.rules [hunk 3]
```diff
@@ -2,16 +2,48 @@ rules_version = '2';
...
+      }
+      match /following/{docId} {
+        allow read: if signedIn();
```
Comment (AngeB21) [inline firebase/firestore/firestore.rules]: maybe you should add a line `allow write: if false;` to explicitly specify that the client should not write.

**Review by AngeB21 — CHANGES_REQUESTED** [review]
Your code is very good🍂! I have only very few requests before approving:

- In **ProfileRepositoryFirebase**, add a `Log.e` if the snapshot listener returns an error for visibility.
- In **FollowListViewModel**, you could consider adding a helper function to check `uid == null` to improve readability.
- In **Firestore rules**, you could explicit the fact that the client shouldn't write.

**Comment by AngeB21** [comment]
Regarding the `firestore.rules`, I suggested adding `allow write if false` for your code clarity, even if it doesn't change the behavior.
